### PR TITLE
impl: symmetry — 6-2-1 net learns 1:2:4 anti-symmetric weights

### LINKED ISSUE #5: impl: symmetry — 6-2-1 net learns 1:2:4 weight ratio
## Goal

Implement the `symmetry` stub following the minimum spec — see #1.

## Source

Rumelhart, Hinton & Williams (1986), "Learning internal representations by error propagation", PDP Vol. 1, Ch. 8.

**Demonstrates:** The 6-2-1 net learns a unique symmetric-but-opposite-sign weight pattern with **1:2:4 magnitude ratio**. Famous "more elegant than the human designers anticipated" result.

Output 1 iff input is a palindrome (symmetric about its midpoint). All 64 6-bit patterns enumerated; 8 are palindromes.

## Reference implementations

- [`encoder-4-2-4/`](../tree/main/encoder-4-2-4) — this repo's worked example
- [`xor/`](../tree/main/xor) — sibling tier-A example, same era, same algorithm (PR #3)
- [`wip-boltzmann-shifter/`](https://github.com/cybertronai/sutro-problems/tree/main/wip-boltzmann-shifter)

## Architecture (from existing stub)

- 6 inputs, 2 hidden, 1 output
- Default lr=0.3, n_sweeps=5000

## The famous result you're trying to reproduce

After convergence, the weight matrix should show:

- **Symmetric magnitudes** about the midpoint (`|w_1| = |w_6|`, `|w_2| = |w_5|`, `|w_3| = |w_4|`)
- **Opposite signs** between mirror-pair positions
- **1:2:4 magnitude ratio** across the three position pairs

The PR is not done unless this property is verified and visualized.

## Files to create in `symmetry/`

Per spec (#1):

- `symmetry.py` (or keep `problem.py`) — dataset + model + train + eval + `inspect_weight_symmetry()` checker
- `README.md` — 8-section template (#1)
- `make_symmetry_gif.py`
- `visualize_symmetry.py` — must include the weight-pattern viz showing the 1:2:4 ratio (e.g., bar chart of `|w_i|` per input position) and a side-by-side comparison to "what the paper described"
- `symmetry.gif`
- `viz/`

## Acceptance checklist

**Original (from #1):**

- [ ] Reproduces in <5 min on a laptop
- [ ] Final accuracy (should be 100% on all 64 patterns) with seed in Results table
- [ ] GIF illustrates problem AND learning dynamics
- [ ] **Weight-symmetry visualization showing the 1:2:4 ratio** — the whole point
- [ ] Training-curves plot
- [ ] "Deviations from the original" section
- [ ] At least one open question
- [ ] No `NotImplementedError` left

**v1 metrics (from #1 comment):**

- [ ] PR body: "Paper reports 1:2:4 ratio with opposite signs; we got <ratio>. Reproduces: yes/no."
- [ ] PR body: wallclock to run the final experiment
- [ ] PR body: implementation wallclock

## Branch & PR

- Branch: `impl/symmetry` off `main`
- PR title: `impl: symmetry`

---

_agent-0bserver07 (Claude Code) on behalf of Yad_

### PR DESCRIPTION
**Pull request by 0bserver07**
## Summary

Implements the `symmetry/` stub per the v1 spec (#1). Closes #5.

Faithful numpy reproduction of Rumelhart, Hinton & Williams (1986),
*"Learning representations by back-propagating errors"* (Nature 323):
6-2-1 sigmoid MLP trained by full-batch backprop with momentum on all
64 6-bit patterns (8 palindromes, 56 non-palindromes). Same convergence
criterion as the paper (`|o − y| < 0.5` for all patterns).

- `symmetry/symmetry.py` — dataset + 6-2-1 MLP + backprop + eval +
  `inspect_weight_symmetry()` + multi-seed sweep + CLI
- `symmetry/visualize_symmetry.py` — training curves + Hinton-diagram
  weights + bar chart of `|w_i|` per input position + observed-vs-paper
  1:2:4 panel + hidden activations across all 64 patterns
- `symmetry/make_symmetry_gif.py` + `symmetry/symmetry.gif` (1.6 MB) —
  animated GIF showing the 1:2:4 anti-symmetric pattern emerging
- `symmetry/README.md` — full 8-section template

## v1 metrics

| | |
|---|---|
| **Reproduces paper?** | **yes**. Paper reports a 1:2:4 magnitude ratio with opposite signs across the three position pairs; we got `1 : 1.994 : 3.969` with anti-symmetry residual `0.000`. Hidden-unit weights at seed 1: `[-1.90, +3.79, -7.55, +7.55, -3.79, +1.90]`. The two hidden units mirror each other in global sign. |
| **Run wallclock** (seed 1) | **0.4 s** for `python3 symmetry.py --seed 1` (1061 sweeps to converge, M-series laptop). 0.65 s wall including process startup per `time(1)`. |
| **Implementation wallclock** | **~12.8 min** end-to-end (start of session → branch pushed). |
| **Final accuracy** | 100% (64/64) for seed 1; 20/30 random seeds converge to 100% within 5000 sweeps; 17/30 of those reach the textbook 1:2:4 / opposite-sign pattern. |

## Reproduces

```
$ time python3 symmetry/symmetry.py --seed 1
...
final accuracy : 100.0% (64/64)
final loss     : 0.0002
converged sweep: 1061
wallclock      : 0.425s
=== weight-symmetry check (W1 = hidden weights, 6 inputs) ===
  hidden 0:
    w           = ['-1.90', '+3.79', '-7.55', '+7.55', '-3.79', '+1.90']
    pair mags   = outer 1.90, middle 3.79, inner 7.55
    sorted-mag ratio = 1 : 1.99 : 3.97
  hidden 1:
    w           = ['+1.88', '-3.74', '+7.45', '-7.45', '+3.74', '-1.88']
    sorted-mag ratio = 1 : 1.99 : 3.97
  sorted-mag ratio (mean over hidden units) = 1 : 1.994 : 3.969  (paper says 1 : 2 : 4)
  anti-symmetry residual (max) = 0.000  (0 = perfectly opposite signs)
  matches the famous 1:2:4 / opposite-sign result? True
```

**Paper reports 1:2:4 ratio with opposite signs; we got `1 : 1.994 : 3.969`. Reproduces: yes.**

## Acceptance checklist

- [x] Reproduces in <5 min on a laptop (~0.4 s actually, full pipeline ~20 s)
- [x] Final accuracy with seed in Results table
- [x] GIF illustrates problem AND learning dynamics (animated 1:2:4 emergence)
- [x] **Weight-symmetry visualization showing the 1:2:4 ratio** — both `viz/weights.png` (Hinton + bar chart) and `viz/weight_pattern.png` (sorted vs paper-prediction comparison)
- [x] Training-curves plot
- [x] "Deviations from the original" section
- [x] Open questions / next experiments (5 entries)
- [x] No `NotImplementedError` left
- [x] PR body: "Paper reports X; we got Y. Reproduces: yes/no."
- [x] PR body: wallclock to run the final experiment
- [x] PR body: implementation wallclock

## Test plan

- [x] `python3 symmetry/symmetry.py --seed 1` — converges at sweep 1061 to 100%
- [x] `python3 symmetry/symmetry.py --seed 7` — converges at sweep 1334 to 100% with same 1:2:4
- [x] `python3 symmetry/symmetry.py --multi-seed 30 --sweeps 5000` — 20/30 converge, 17/30 match 1:2:4 (12.7 s total)
- [x] `python3 symmetry/visualize_symmetry.py --seed 1` — 4 PNGs in `symmetry/viz/`
- [x] `python3 symmetry/make_symmetry_gif.py --seed 1` — `symmetry/symmetry.gif` (1.6 MB)

---
_agent-0bserver07 (Claude Code) on behalf of Yad_

### CODE REVIEW

**Comment by 0bserver07** [comment]
## Automated review — verdict: **MERGE**

Read-only audit by `@_agent-0bserver07`. All checks pass against the v1 spec (#1).

| Check | Status |
|---|---|
| Required files | ✓ all present |
| README 8-section template | ✓ complete |
| Running command verified | ✓ `python3 symmetry.py --seed 1` executes in <0.4s, converges at epoch 1061 |
| Results table | ✓ seed 1, encoding={-1,+1}, lr=0.3, momentum=0.95, init_scale=1.0; sorted ratio + anti-symmetry residual table |
| GIF size | ✓ 1.6 MB |
| v1 metrics in PR body | ✓ all 3 (Reproduces: yes / Run wallclock 0.4 s / Implementation wallclock 12.8 min) |
| Files outside `symmetry/` | ✓ none |
| `NotImplementedError` remaining | ✓ none |

**Headline result reproduced cleanly**: observed ratio `1 : 1.994 : 3.969` with anti-symmetry residual 0.000. 17/30 seeds land on the canonical 1:2:4 invariant; 20/30 reach 100% accuracy. Includes `inspect_weight_symmetry()` analysis function that quantifies the result numerically.

Approved by automated audit. Awaiting human merge per repo policy.

---

_agent-0bserver07 (Claude Code) on behalf of Yad_

**Comment by 0bserver07** [comment]
Superseded by wave-level PR #32.
